tests/test_resolver.py

```diff
@@ -72,48 +93,144 @@ def test_resolver():
     assert resolver.get_judgement(a_canon, "a1") == Judgement.NO_JUDGEMENT
     assert resolver.get_judgement("b1", "b2") == Judgement.POSITIVE
 
+    # Can we actually commit after all these operations?
+    resolver.commit()
+
+
+def test_cluster_to_cluster():
+    resolver = Resolver.make_default()
+    resolver.begin()
+
+    a_canon = resolver.decide("a1", "a2", Judgement.POSITIVE)
+    b_canon = resolver.decide("b1", "b2", Judgement.POSITIVE)
+    resolver.decide(a_canon, b_canon, Judgement.UNSURE)
+    resolver.decide(a_canon, "a3", Judgement.POSITIVE)
+    resolver.remove("a3")
+
+    assert "a1" in resolver.connected(Identifier.get("a1"))
+    assert "a2" in resolver.connected(Identifier.get("a1"))
+    assert "b1" not in resolver.connected(Identifier.get("a1"))
+    assert Edge(a_canon, b_canon) in set(resolver._get_resolved_edges("a1", "b1"))
+
+    # ab_canon = resolver.decide("a1", "b1", Judgement.POSITIVE)
+    # TODO: There's a bug here - decide(a, b, POSITIVE) must always return a canonical.
+    # assert ab_canon.canonical, ab_canon
```

Comment:
TODO

Reply:
the bug occurs when a1 and b1 already have a canonical.

Should it be a requirement that the arguments to decide are the result of resolver.canonical()? Should we make decide() do that for us?

Reply:
get the canonical of of the arguments, handle recursion